File: 06. Backtracking/03. Palindromic Partitioning/main.py
# Function to check if a substring is a palindrome
def is_palindrome(s, start, end):
    while start < end:
        if s[start] != s[end]:
            return False
        start += 1
        end -= 1
    return True

# Backtracking function to generate partitions
def backtrack(start, s, current, result):
    if start == len(s):
        result.append(list(current))
        return
    
    for end in range(start, len(s)):
        if is_palindrome(s, start, end):
            current.append(s[start:end + 1])  # Take substring
            backtrack(end + 1, s, current, result)
            current.pop()  # Backtrack

# Function to generate all palindromic partitions
def partition(s):
    result = []
    backtrack(0, s, [], result)
    return result

# Utility function to print partitions
def print_partitions(partitions):
    print("[")
    for partition in partitions:
        print("  ", partition)
    print("]")

# Main function
def main():
    s = "aab"
    print(f'Palindromic Partitions of "{s}":')
    print_partitions(partition(s))

main()

File: 06. Backtracking/03. Palindromic Partitioning/main.js
// Function to check if a substring is a palindrome
function isPalindrome(s, start, end) {
    while (start < end) {
        if (s[start] !== s[end]) return false;
        start++;
        end--;
    }
    return true;
}

// Backtracking function to generate partitions
function backtrack(start, s, current, result) {
    if (start === s.length) {
        result.push([...current]);
        return;
    }

    for (let end = start; end < s.length; end++) {
        if (isPalindrome(s, start, end)) {
            current.push(s.substring(start, end + 1)); // Take substring
            backtrack(end + 1, s, current, result);
            current.pop(); // Backtrack
        }
    }
}

// Function to generate all palindromic partitions
function partition(s) {
    let result = [];
    backtrack(0, s, [], result);
    return result;
}

// Utility function to print partitions
function printPartitions(partitions) {
    partitions.forEach(partition => console.log("  ", partition));
}

// Main function
function main() {
    let s = "aab";
    console.log(`Palindromic Partitions of "${s}":`);
    printPartitions(partition(s));
}

main();
